### DISCUSSION TITLE
Tippy v7

### DISCUSSION THREAD

**Discussion by atomiks** [atomiks/tippyjs#991] [Announcements]
# Tippy v7

It’s now been 1.5 years since Tippy 6 released to use the brand new Popper 2 positioning engine, and is the longest-lasting major version of the library to date. As time has passed I have gathered more feedback and ideas for making the library the best it can possibly be.

My main goal with this new major version is to focus on accessibility, API streamlining for renderless styling, modularity, performance, and documentation.

## Accessibility

Tippy has had basic accessibility features since v0 in 2017 and has steadily improved, but the area most notably lacking is with interactive tippies. While proper keyboard management can be added by the developer on top of the library, baked-in accessibility is very important for the web as most developers of small projects will neglect users of assistive technology.

Focus management of interactive content within a tippy is enabled, but it requires appending the tippy element next to the reference element to enable focus to move within the popover immediately. This is done by default, but can cause issues in some cases (when setting `interactive: true`) in terms of positioning and styling. I believe the library should handle focus by default in some way depending on the context the tippy is being used as without appending next to the reference element.

For more complex UI elements, like a dropdown menu, there is no arrow key navigation for instance. Dismissing a tooltip is also not currently possible by default, unless added by the developer.

## API streamlining for renderless styling

Some legacy render/styling-related props were added, such as `inertia`, back when the library focused more on being a fun tooltip library to quickly add tooltips to your website, rather than a more robust, headless type of solution to use in UI libraries like React. These props have no use to headless users for the most part, unless their `render` function utilizes it in some way, resulting in pollution of the `Props` object.

Headless UI libraries are the future, and I want Tippy to move more in this direction. A lot of work was already done to enable this in v6, but the API can be streamlined to better separate `render`-related props and `logic`-related props for users.

Of course, the plug-n-play nature of Tippy is important for adoption, so this will remain a focus too.

## Modularity

Users of libraries will find it annoying when their use-case is not supported by the library. Tippy made strides with this in v6 by creating a plugin system to extend the library, but a ton of the functionality is still baked-in to core, and therefore cannot be treeshaken away by bundlers. Further, it's difficult to modify behavior of existing props, such as `interactive`. Splitting out more of the core into plugins should enable more flexibility for users. The idea behind improving this area more is to retain the "library doesn't support my use case" users by letting them know their use case can be added on top easily.

## Performance

There have been reports of delays regarding hiding or showing the tippy. This might be related to appending the tippy to the `<body />` when there are a large number of nodes in the user's document. An article recently written showed that these poppers should be appended to a separate `<div />` node within `<body />` to improve performance. Perhaps this should be enforced by default, or at least documented better.

## Documentation

There was a shift of focus in v6 away from tippy being solely a "tooltip" library and more toward generic "popper" elements in general. I believe the documentation can be improved in this regard, especially a `Recipes` section with a ton of examples and use-cases for users to look at beyond just basic tooltips. A cool landing page and a proper domain not hosted on GitHub pages will likely make the library more attractive too.

## Singletons

Singletons were a quick, fun idea I came up with and put on the docs to showcase what you can do with Tippy. Right away, people wanted to know how to make them, so an official API was created thereafter. Over the years however, singletons have proven to be quite complex and problematic in the following ways:

- Frequently had bugs, tons of tricky unexpected edge cases need to be handled
- Were difficult to make accessible if interactive
- Did not transition the size of the tippy element itself for a smoother experience with `moveTransition`
- Prevented making “lazy” behaviour the default in the React wrapper, resulting in a performance cost
- `disabled` and visible` props in React did not work

They should be rewritten with these issues in mind.

## Your thoughts?

What pain points have you experienced using the library, and how do you think it can be improved? I'd like to hear your thoughts. My goal is to get this new version released by 2022.

**Comment by silouanwright** [comment]
* A consistent maintainer. For those reading this, please sponsor him! https://github.com/sponsors/atomiks
* My major pain points:
  * This issue with the react library prevents me from using tippy: https://github.com/atomiks/tippyjs-react/issues/209 (I tried to implement this myself with the following gist but it failed once I started testing with react testing library: https://gist.github.com/atomiks/520f4b0c7b537202a23a3059d4eec908)
  * Performance (delay when opening)
  
I definitely would love to see that increased focus on headless & accessibility!

**Reply by atomiks** [reply]
Thank you so much for sponsoring me! For the React wrapper that's definitely something I want to implement by default for v5. The way singletons were written is what made it unfeasible for the current version based on my inspection.

**Comment by atomiks** [comment]
## Progress Update 1

I've been making progress on my local v7 branch. Once the code is in a decent state I can make a PR if necessary for review, including screenshots of the new docs.

### API changes

- `animation` has been split out into a plugin and now supports an object value with `show` and `hide` functions to implement JavaScript/spring animations with the default template, e.g. with Popmotion.
- `role` has been moved into the `aria` object, which has now also become a plugin
- animation-related properties `inertia` and `duration` removed to reduce Props pollution for headless usage, as they can be set via CSS easily
- general code cleanup

### Exploration

**Modularity:** I've been experimenting with making virtually everything a plugin but quickly realized that certain props are deeply linked to the behavior of other ones, making it difficult to split them out. If they provided data in some form, like Popper has with `modifiersData`, it could be possible but I'll need to spend a lot of time experimenting and I'm not sure if it'd be worth it, or to keep them in core/opinionated.

**Accessibility:** I'm thinking `interactive: true` should not append to the reference's parentNode by default (due to unexpected and hard to debug issues) and instead create plugins for keyboard navigation depending on the context in which the tippy is being used. This would include focus trapping for popovers, arrow key navigation for dropdown menu items, etc. The most important part here will be the documentation to ensure the developer follows a11y guidelines. It's difficult to keep the `tippy()` function generic otherwise, as we can't assume what the UI component is.

**Performance:** I'm thinking we continue to use `document.body` as the default (as we can't assume anything else) but strongly highlight in the documentation that appending via a separate node within the document is recommended. I believe this may be the source of performance issues https://web.archive.org/web/20210827084020/https://atfzl.com/don-t-attach-tooltips-to-document-body.

### Docs

This will by far take up the most time but I think the payoff will be worth it. I'm completely starting from scratch and taking inspiration from Parcel 2, Tailwind and Apple's landing pages. I'm using Tailwind for CSS styling and Next.js for the static generation, and so far they're both more enjoyable to use than the current docs setup. I bought the `tippyjs.com` domain too, which looks much neater in the URL bar. I plan to include a ton of examples which will be clickable via cards.

**Comment by chasegiunta** [comment]
Hi @atomiks thank you for all the great work you've done on Tippy thus far. I'm using Tippy for for numerous cases (from context menus, dropdowns, to anchored popovers that change arrow direction depending where they are offscreen via popper modifiers). The couple pain points so far for me have been:
- Having to write custom plugins to style the various tippy elements (box, content, arrow). I've seen your code written in a issue/discussion somewhere so I've replicated that and create three plugins to handle these elements, but it would ideal if it was core (or official plugin).
- Not sure if this is even possible, and I could be wrong, but I have to guess that singletons are a large pain point for _you_ to support. Having to create multiple tippy instances for every target, and then a singleton to essentially override the tippy instances, seems like a weird workflow to me just to transition between target/content or modify a delay. I'm just curious if it's possible to emulate the same singleton behavior with one tippy instance, swapping out targets, content, etc. We had some race condition issues with initiating singletons before/after tippy instances with our framework, had to use timeouts, and ultimately ditched singletons to cleanup the code at the cost of a better user experience.

Here's to Tippy v7 🍻 . Thanks again.

**Reply by atomiks** [reply]
Hi @chasegiunta thank you for your feedback, very appreciated!

- Regarding this first dot point, are you able to provide some examples of what you're doing with those elements?
- You're right, I think `createSingleton` needs to be completely rewritten based on the feedback over the years. There have been many various bugs when using them, one just recently opened here; and in React, they prevented making "lazy" behavior default, which a lot of people wanted. I have ideas to make them better and more accessible without the pitfalls.

**Reply by chasegiunta** [reply]
@atomiks Sure! And I guess I should elaborate on that point - we're trying to get away from writing stylesheets and using Tailwind for everything, hence the need for the plugins. I'm aware of the themes usage, but in an attempt to keep everything out of css, I'm using plugins so we directly hand off props like `@arrowClass="bg-sky-500"` to our tippy components. I imagine others may be in that boat, but if that ask is an overkill, I get that, too.
https://user-images.githubusercontent.com/1377169/140397740-2687df78-df63-4ddc-a863-7f6dfaecb5af.mp4

**Reply by atomiks** [reply]
For Tailwind, I am using the `render` function in React to create my tippy element and therefore apply the classNames to style it in the new documentation website I am building for v7. 

In vanilla tippy, you can also use the `render` function to construct the tippy's DOM from scratch and apply Tailwind's class names that way, but, it's not particularly DX-friendly. This type of thing is another important factor in favor of the headless direction.

**Comment by atomiks** [comment]
# Progress Update 2

## Concept rebrand

I've decided to rebrand the library to be more generic and highlight the move to a more "low-level" nature

The new slogan is:

> Floating UI elements for the web

instead of:

> Tooltip, popover, dropdown, and menu library

It's simpler and more tersely describes the library. The old slogan sort of implied that it could only craft those 4, and that there were "prebuilt" configurations e.g. (`tooltip()`, `popover()`, `dropdown()` functions instead of one generic `tippy()`). One potential downside to this is SEO, but I'm unsure the extent of that yet.

The dependency list is like: `Popper` is used by `Tippy` to position itself, and `Tippy` is used to build your `Tooltip`, `Popover`, `Dropdown`, and more. 

For instance, Bootstrap's tooltips and popovers the highest level you can achieve, since they are pre-styled, pre-rendered, and the functionality is pre-built. Popper is the lowest level you can achieve, as it just positions an element. Tippy exists as sort of an in-between abstraction for developers to use.

## Landing page

Here are some sections I've made:

<img width="1377" alt="Main section showing 3 subsections: tooltips, popovers, and dropdowns, with an interactive example next to each" src="https://user-images.githubusercontent.com/22450188/140939347-b1fe8a57-b5a6-4dff-bf27-2275366cc210.png">
<img width="1351" alt="Main section showing a scrollable container with a floating box inside to showcase Popper's positioning engine" src="https://user-images.githubusercontent.com/22450188/140939477-317a18a0-cab5-4f52-9af8-f390306db423.png">

## Accessibility

Because `tippy()` is generic and low-level, we can't assume the UI element being created, nor can we provide focus management out of the box. I've even decided to require importing `aria` (for ESM), to allow the developer to craft their own `aria` solution if necessary. 

Instead, I believe that good documentation will be the key to enabling developers using Tippy to make their element accessible. 

The `Accessibility` section therefore will be a key part and linked frequently, **especially** for the `interactive` prop.

<img width="836" alt="Tippy accessibility docs showing the usage of the `aria` plugin to make non-interactive tippys accessible to screen readers" src="https://user-images.githubusercontent.com/22450188/140940867-3b70fc64-f8bb-4302-adf5-95a1ac8fc43d.png">

e.g. for interactive popovers, I have created a section demonstrating focus trapping or DOM source order techniques of keyboard navigation.

The key takeaway is that Tippy is a low-level builder for all types of floating elements, and **not** a higher-level preconfigured UI element builder, like a `tooltip()` or `popover()` function would be (or `<Tooltip />` in React, etc). Therefore being "accessible without config or documentation" isn't possible, just like plain HTML.

### More API changes

- In TypeScript, I've made `instance.state` extensible to store custom data inside.
- Plugins can now return a `props` object if they want to set some props on initialization (instead of `onCreate` + `setProps`)
- I'm moving as many props as I can to a plugin for tree-shaking and configurability purposes. By doing so I am discovering more things that need to be exposed for plugin developers which is improving the extensibility of the library.
- `onClickOutside` is now `onDismiss` and I'll need to explore changes to `hideOnClick` given the concept of "dismissability" encompassing `esc`, etc.

**Comment by atomiks** [comment]
So small update: while working on this I also got some ideas for an API rewrite for Popper v3 (Tippy's dependency) which I have put all my focus on for now. It's going to be far smaller in weight, platform-agnostic (can run on React Native), easier to extend, etc. Since Tippy is built on it, it will also improve Tippy's own usage/API, making all the ideas I had for v7 work better.

The rewrite is almost done, with only a few things remaining, then I can put all focus back onto this 😃 

- Dicussion: https://github.com/popperjs/popper-core/discussions/1415
- PR: https://github.com/popperjs/popper-core/pull/1416

Been posting about it on Twitter: https://twitter.com/atomiksdev

**Comment by atomiks** [comment]
The concepts I had for Tippy v7 are now going to be created under my new `Floating UI` project, under a `@floating-ui/components` package. 

Tippy.js is going to be in lower maintenance mode from this point. I created it in early 2017 in like 3 days when I had no idea what I was doing, and somehow it got 2K stars in a couple of days when I released it. But I've learned so much since then thanks to this library and other open source projects, that it's better to start something new with all this new knowledge than to refactor all the legacy baggage.

- Floating UI: https://floating-ui.com
- Dicussion: https://github.com/atomiks/floating-ui/discussions/18

❤️
